### PR TITLE
[MM-405]: Fixed the pull review request events not working

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
#### Summary
Fixed the issue that the pull request review event not working. Removed the GetComment.GetDiffHunk from the template as it can be very long and not that very useful. Also changed the name for the type of the post. 

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-github/issues/405

### CODE REVIEW

File: server/plugin/webhook.go [hunk 1]
```diff
@@ -937,7 +937,7 @@ func (p *Plugin) postPullRequestReviewCommentEvent(event *github.PullRequestRevi
 
 	post := &model.Post{
 		UserId:  p.BotUserID,
-		Type:    "custom_git_pull_review_comment",
+		Type:    "custom_git_pr_comment",
```
Comment (hanzei) [inline server/plugin/webhook.go:940]: While we are not using the custom type currently, I think there is value in keeping the type the same for each post.

What is the reason behind changing it?

File: server/plugin/template.go [hunk 2]
```diff
@@ -303,9 +303,9 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 
 	template.Must(masterTemplate.New("pullRequestReviewEvent").Funcs(funcMap).Parse(`
 {{template "repo" .GetRepo}} {{template "user" .GetSender}}
-{{- if eq .GetReview.GetState "APPROVED"}} approved
-{{- else if eq .GetReview.GetState "COMMENTED"}} commented on
-{{- else if eq .GetReview.GetState "CHANGES_REQUESTED"}} requested changes on
```
Comment (hanzei) [inline server/plugin/template.go:308]: Did GitHub document the changes anywhere?

**Review by hanzei — CHANGES_REQUESTED** [review]
Glad to see this bug getting fixed :rocket:

File: server/plugin/webhook.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/plugin/webhook.go:940]: I kept getting the error that the length of the type should be less than 2, that's why I have reduced it. 
@hanzei

File: server/plugin/template.go (hunk 2 above)
Comment (Kshitij-Katiyar) [reply server/plugin/template.go:308]: @hanzei I was getting the state in small case and in the code as well they have been changed to lower case in the using toLower. Let me try to find the GitHub documentation for this.

File: server/plugin/template.go [hunk 3]
```diff
@@ -314,7 +314,6 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 	template.Must(masterTemplate.New("newReviewComment").Funcs(funcMap).Parse(`
 {{template "repo" .GetRepo}} New review comment by {{template "user" .GetSender}} on {{template "pullRequest" .GetPullRequest}}:
 
-{{.GetComment.GetDiffHunk}}
```
Comment (mickmister) [inline server/plugin/template.go:317]: Can you provide a screenshot for what this looks like?

**Review by mickmister — COMMENTED** [review]
@Kshitij-Katiyar There's a PR that was submitted that solves this same issue https://github.com/mattermost/mattermost-plugin-github/pull/778. Can you please take a look there and compare any changes that may need to be added here?

File: server/plugin/template.go (hunk 2 above)
Comment (Kshitij-Katiyar) [reply server/plugin/template.go:308]: @hanzei Can't find a documentation but here is a screenshot of the webhook payload
Screenshot from 2024-06-21 12-33-38

**Comment by Kshitij-Katiyar** [comment]
> @Kshitij-Katiyar There's a PR that was submitted that solves this same issue #778. Can you please take a look there and compare any changes that may need to be added here?

Thanks for the PR, added some useful changes from the PR

File: server/plugin/template.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/plugin/template.go:317]: **Old UI**
Screenshot from 2024-06-21 12-46-20

**New UI**
new

File: webapp/package.json [hunk 4]
```diff
@@ -58,7 +58,7 @@
     "sass": "1.60.0",
     "sass-loader": "10.4.1",
     "style-loader": "1.2.1",
-    "typescript": "4.0.3",
+    "typescript": "4.2.3",
```
Comment (mickmister) [inline webapp/package.json:61]: This is adding a lot to the diff of the PR

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 5]
```diff
@@ -350,7 +350,7 @@ function getReviewText(item: GithubItem, style: any, secondLine: boolean) {
             return false;
         }
 
-        if (v.state === 'COMMENTED' || v.state === 'DISMISSED') {
+        if (v.state === 'commented' || v.state === 'dismissed') {
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:353]: Can we put these into a typescript enum?

File: server/plugin/webhook.go [hunk 6]
```diff
@@ -908,6 +908,19 @@ func (p *Plugin) postPullRequestReviewCommentEvent(event *github.PullRequestRevi
 		return
 	}
 
+	post := &model.Post{
```
Comment (mickmister) [inline server/plugin/webhook.go:911]: Putting the post above the loop reintroduces the bug that https://github.com/mattermost/mattermost-plugin-github/pull/755 solves. Can we revert the changes here that are not related to `approved` etc.?

File: server/plugin/webhook.go [hunk 7]
```diff
@@ -908,6 +908,19 @@ func (p *Plugin) postPullRequestReviewCommentEvent(event *github.PullRequestRevi
...
+		UserId:  p.BotUserID,
+		Type:    "custom_git_pr_comment",
+		Message: newReviewMessage,
+	}
+
+	repoName := strings.ToLower(repo.GetFullName())
+	commentID := event.GetComment().GetID()
```
Comment (mickmister) [inline server/plugin/webhook.go:918]: Note that these are quick functions that do not do network calls, in case this was moved up here for performance reasons

File: server/plugin/webhook.go [hunk 8]
```diff
@@ -935,15 +948,6 @@ func (p *Plugin) postPullRequestReviewCommentEvent(event *github.PullRequestRevi
...
-
-		repoName := strings.ToLower(repo.GetFullName())
-		commentID := event.GetComment().GetID()
-
-		post.AddProp(postPropGithubRepo, repoName)
-		post.AddProp(postPropGithubObjectID, commentID)
-		post.AddProp(postPropGithubObjectType, githubObjectTypePRReviewComment)
-
 		post.ChannelId = sub.ChannelID
 		if err = p.client.Post.CreatePost(post); err != nil {
```
Comment (mickmister) [inline server/plugin/webhook.go:952]: Referencing https://github.com/mattermost/mattermost-plugin-github/pull/768/files#r1649049447 the `p.client.Post.CreatePost(post)` here mutates the post instead of returning the created post, which has had side effects in this plugin. I think we should instead use `p.API.CreatePost(post)` to avoid the risk of introducing the issue on unrelated code changes cc @hanzei

File: webapp/package.json (hunk 4 above)
Comment (Kshitij-Katiyar) [reply webapp/package.json:61]: @mickmister Without these changes the `ci-lint` will fail. Let me know if you want us to find any other way around to fix the failing lint test other than this.

**Comment by Kshitij-Katiyar** [comment]
@mickmister Fixed the comments, please look into it.

File: server/plugin/webhook.go [hunk 9]
```diff
@@ -945,8 +945,8 @@ func (p *Plugin) postPullRequestReviewCommentEvent(event *github.PullRequestRevi
 		post.AddProp(postPropGithubObjectType, githubObjectTypePRReviewComment)
 
 		post.ChannelId = sub.ChannelID
-		if err = p.client.Post.CreatePost(post); err != nil {
-			p.client.Log.Warn("Error webhook post", "post", post, "error", err.Error())
+		if _, appErr := p.API.CreatePost(post); appErr != nil {
+			p.client.Log.Warn("Error webhook post", "post", post, "error", appErr.Error())
```
Comment (hanzei) [inline server/plugin/webhook.go:949]: `p.API` should not be used. The returned `model.AppError` makes things way too hard. `p.client` is they way to go.

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go (hunk 9 above)
Comment (raghavaggarwal2308) [reply server/plugin/webhook.go:949]: @hanzei Please refer to this comment: https://github.com/mattermost/mattermost-plugin-github/pull/768#discussion_r1649367568

Comment (mickmister) [reply server/plugin/webhook.go:949]: In this case, the `p.client.Post` method has essentially a bug in it, that has caused issues in this plugin because it mutates the post that is passed in. Is it really a "bug"? Maybe not, but I think we should avoid using it since it has caused issues in production that we then need to fix

2/5 The benefits of using `error` over `AppError` in this case are not worth it imo. I think keeping it as `p.client.Post` will make potential issues come back, just as they did in this PR

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 10]
```diff
@@ -19,6 +19,7 @@ import TickIcon from '../../images/icons/tick';
 import SignIcon from '../../images/icons/sign';
 import ChangesRequestedIcon from '../../images/icons/changes_requested';
 import {getLabelFontColor} from '../../utils/styles';
+import {ReviewState} from '../../constants/index';
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:22]:
```suggestion
import {ReviewState} from '../../constants';
```

File: server/plugin/webhook.go (hunk 9 above)
Comment (mickmister) [reply server/plugin/webhook.go:949]: @Kshitij-Katiyar I'm thinking we should keep the changes in this PR to a minimum so let's keep this as-is

**Comment by Kshitij-Katiyar** [comment]
@mickmister Fixed the comments, please have a look

**Review by mickmister — APPROVED** [review]

**Review by hanzei — APPROVED** [review]

**Review by AayushChaudhary0001 — APPROVED** [review]
The above PR has been tested for the following scenarios:

- Checked the pull review request events
- Tested the notification for pull review request events 

The PR was working fine for all the conditions, LGTM. Approved
